File: src/lexer/symbol.rs
/// 記号・演算子トークン解析メソッド。
///
/// `Lexer` の実装を機能単位に分割した補助ファイル。
/// メソッドは `scan.rs` で定義された `Lexer` に対する `impl` ブロックとして提供する。

use crate::token::Token;

use super::scan::Lexer;

impl Lexer {
    /// 記号文字から演算子・区切り記号トークンを生成する。
    ///
    /// 先頭文字を消費したうえで、複数文字からなるトークン（`==`, `->`, `**=` など）は
    /// 1 文字先読みをして確定させる。括弧文字では `bracket_depth` の更新も行う。
    ///
    /// # 戻り値
    /// 対応する `Token`。未知の文字の場合は `Token::Unknown(char)`
    pub(super) fn lex_symbol(&mut self) -> Token {
        let ch = self.bump().unwrap();
        match ch {
            // 開き括弧：bracket_depth をインクリメントする
            '(' => {
                self.bracket_depth += 1;
                Token::LParen
            }
            // 閉じ括弧：bracket_depth が 0 を下回らないようにデクリメントする
            ')' => {
                if self.bracket_depth > 0 {
                    self.bracket_depth -= 1;
                }
                Token::RParen
            }
            '[' => {
                self.bracket_depth += 1;
                Token::LBracket
            }
            ']' => {
                if self.bracket_depth > 0 {
                    self.bracket_depth -= 1;
                }
                Token::RBracket
            }
            '{' => {
                self.bracket_depth += 1;
                Token::LBrace
            }
            '}' => {
                if self.bracket_depth > 0 {
                    self.bracket_depth -= 1;
                }
                Token::RBrace
            }

            // `+` または `+=`
            '+' => {
                if self.ch() == Some('=') {
                    self.pos += 1;
                    Token::PlusEq
                } else {
                    Token::Plus
                }
            }

            // `-` / `-=` / `->` の 3 種を判定する
            '-' => match self.ch() {
                Some('=') => {
                    self.pos += 1;
                    Token::MinusEq
                }
                Some('>') => {
                    self.pos += 1;
                    Token::Arrow
                }
                _ => Token::Minus,
            },

            // `*` / `*=` / `**` / `**=` の 4 種を判定する
            '*' => match self.ch() {
                Some('*') => {
                    self.pos += 1;
                    if self.ch() == Some('=') {
                        self.pos += 1;
                        Token::StarStarEq
                    } else {
                        Token::StarStar
                    }
                }
                Some('=') => {
                    self.pos += 1;
                    Token::StarEq
                }
                _ => Token::Star,
            },

            // `/` / `/=` / `//` / `//=` の 4 種を判定する
            '/' => match self.ch() {
                Some('/') => {
                    self.pos += 1;
                    if self.ch() == Some('=') {
                        self.pos += 1;
                        Token::SlashSlashEq
                    } else {
                        Token::SlashSlash
                    }
                }
                Some('=') => {
                    self.pos += 1;
                    Token::SlashEq
                }
                _ => Token::Slash,
            },

            // `%` または `%=`
            '%' => {
                if self.ch() == Some('=') {
                    self.pos += 1;
                    Token::PercentEq
                } else {
                    Token::Percent
                }
            }

            // `@` または `@=`
            '@' => {
                if self.ch() == Some('=') {
                    self.pos += 1;
                    Token::AtEq
                } else {
                    Token::At
                }
            }

            // `=` / `==` / `=>`
            '=' => {
                if self.ch() == Some('=') {
                    self.pos += 1;
                    Token::EqEq
                } else if self.ch() == Some('>') {
                    self.pos += 1;
                    Token::FatArrow
                } else {
                    Token::Eq
                }
            }

            // `!=`（`!` 単独は未知文字）
            '!' => {
                if self.ch() == Some('=') {
                    self.pos += 1;
                    Token::NotEq
                } else {
                    Token::Unknown('!')
                }
            }

            // `<` / `<=` / `<<` / `<<=` / `<-` の 5 種を判定する
            '<' => match self.ch() {
                Some('<') => {
                    self.pos += 1;
                    if self.ch() == Some('=') {
                        self.pos += 1;
                        Token::LtLtEq
                    } else {
                        Token::LtLt
                    }
                }
                Some('=') => {
                    self.pos += 1;
                    Token::LtEq
                }
                Some('-') => {
                    self.pos += 1;
                    Token::LeftArrow
                }
                _ => Token::Lt,
            },

            // `>` / `>=` / `>>` / `>>=` の 4 種を判定する
            '>' => match self.ch() {
                Some('>') => {
                    self.pos += 1;
                    if self.ch() == Some('=') {
                        self.pos += 1;
                        Token::GtGtEq
                    } else {
                        Token::GtGt
                    }
                }
                Some('=') => {
                    self.pos += 1;
                    Token::GtEq
                }
                _ => Token::Gt,
            },

            // `&` または `&=`
            '&' => {
                if self.ch() == Some('=') {
                    self.pos += 1;
                    Token::AmpEq
                } else {
                    Token::Amp
                }
            }

            // `|` または `|=`
            '|' => {
                if self.ch() == Some('=') {
                    self.pos += 1;
                    Token::PipeEq
                } else {
                    Token::Pipe
                }
            }

            // `^` または `^=`
            '^' => {
                if self.ch() == Some('=') {
                    self.pos += 1;
                    Token::CaretEq
                } else {
                    Token::Caret
                }
            }

            // `~`（単独のみ）
            '~' => Token::Tilde,

            // `:` / `::` / `:=` の 3 種を判定する
            ':' => {
                if self.ch() == Some(':') {
                    self.pos += 1;
                    Token::ColonColon
                } else if self.ch() == Some('=') {
                    self.pos += 1;
                    Token::ColonEq
                } else {
                    Token::Colon
                }
            }

            ',' => Token::Comma,
            ';' => Token::Semicolon,

            // `.` または `...`（`..` は存在しない）
            '.' => {
                if self.ch() == Some('.') && self.ch1() == Some('.') {
                    self.pos += 2;
                    Token::Ellipsis
                } else {
                    Token::Dot
                }
            }

            // 上記以外はすべて未知文字として返す
            other => Token::Unknown(other),
        }
    }
}

File: impl_python/lexer/symbol.py
# git SHA: aea2e1fe6909a7aed9643a2e7184f19fd0195ccc
"""Symbol and operator tokenization (mirrors src/lexer/symbol.rs)."""
from __future__ import annotations
from ..token import Token, TokenKind


class _LexerSymbol:
    """Mixin providing symbol/operator tokenization."""

    def _lex_symbol(self) -> Token:
        c = self._bump()
        assert c is not None

        if c == '(':
            self._bracket_depth += 1
            return Token(TokenKind.LPAREN)
        if c == ')':
            if self._bracket_depth > 0:
                self._bracket_depth -= 1
            return Token(TokenKind.RPAREN)
        if c == '[':
            self._bracket_depth += 1
            return Token(TokenKind.LBRACKET)
        if c == ']':
            if self._bracket_depth > 0:
                self._bracket_depth -= 1
            return Token(TokenKind.RBRACKET)
        if c == '{':
            self._bracket_depth += 1
            return Token(TokenKind.LBRACE)
        if c == '}':
            if self._bracket_depth > 0:
                self._bracket_depth -= 1
            return Token(TokenKind.RBRACE)

        if c == '+':
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.PLUS_EQ)
            return Token(TokenKind.PLUS)

        if c == '-':
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.MINUS_EQ)
            if self._ch() == '>':
                self._pos += 1; return Token(TokenKind.ARROW)
            return Token(TokenKind.MINUS)

        if c == '*':
            if self._ch() == '*':
                self._pos += 1
                if self._ch() == '=':
                    self._pos += 1; return Token(TokenKind.STAR_STAR_EQ)
                return Token(TokenKind.STAR_STAR)
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.STAR_EQ)
            return Token(TokenKind.STAR)

        if c == '/':
            if self._ch() == '/':
                self._pos += 1
                if self._ch() == '=':
                    self._pos += 1; return Token(TokenKind.SLASH_SLASH_EQ)
                return Token(TokenKind.SLASH_SLASH)
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.SLASH_EQ)
            return Token(TokenKind.SLASH)

        if c == '%':
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.PERCENT_EQ)
            return Token(TokenKind.PERCENT)

        if c == '@':
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.AT_EQ)
            return Token(TokenKind.AT)

        if c == '=':
            if self._ch() == '=':
                self._pos += 1
                if self._ch() == '=':
                    self._pos += 1; return Token(TokenKind.EQ_EQ_EQ)
                return Token(TokenKind.EQ_EQ)
            if self._ch() == '>':
                self._pos += 1; return Token(TokenKind.FAT_ARROW)
            return Token(TokenKind.EQ)

        if c == '!':
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.NOT_EQ)
            return Token(TokenKind.UNKNOWN, value='!')

        if c == '<':
            if self._ch() == '<':
                self._pos += 1
                if self._ch() == '=':
                    self._pos += 1; return Token(TokenKind.LT_LT_EQ)
                return Token(TokenKind.LT_LT)
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.LT_EQ)
            # <- is not a single token; emit LT, leave '-' for next call
            return Token(TokenKind.LT)

        if c == '>':
            if self._ch() == '>':
                self._pos += 1
                if self._ch() == '=':
                    self._pos += 1; return Token(TokenKind.GT_GT_EQ)
                return Token(TokenKind.GT_GT)
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.GT_EQ)
            return Token(TokenKind.GT)

        if c == '&':
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.AMP_EQ)
            return Token(TokenKind.AMP)

        if c == '|':
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.PIPE_EQ)
            return Token(TokenKind.PIPE)

        if c == '^':
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.CARET_EQ)
            return Token(TokenKind.CARET)

        if c == '~':
            return Token(TokenKind.TILDE)

        if c == ':':
            if self._ch() == ':':
                self._pos += 1; return Token(TokenKind.COLON_COLON)
            if self._ch() == '=':
                self._pos += 1; return Token(TokenKind.COLON_EQ)
            return Token(TokenKind.COLON)

        if c == ',':
            return Token(TokenKind.COMMA)
        if c == ';':
            return Token(TokenKind.SEMICOLON)

        if c == '.':
            if self._ch() == '.' and self._ch1() == '.':
                self._pos += 2; return Token(TokenKind.ELLIPSIS)
            return Token(TokenKind.DOT)

        return Token(TokenKind.UNKNOWN, value=c)
